### DISCUSSION TITLE
multi-node distributed training

### DISCUSSION THREAD

**Discussion by gogogwwb** [hpcaitech/ColossalAI#1124] [Community | Q&A] [answered]
Is there an example of multi-node distributed training?

**Comment by BoxiangW** [comment, marked as answer]
We have provided [ColossalAI-Examples](https://github.com/hpcaitech/ColossalAI-Examples) for various examples regarding our features and training processes for different tasks.

**Reply by gogogwwb** [reply]
Can I use multiple machines and multiple cards to train[ the demo of gpt](https://github.com/hpcaitech/ColossalAI-Examples/tree/main/language/gpt), just modify the startup command?
colossalai run --nproc_per_node=<num_gpus> --host host1,host2 --master_addr host1 train_gpt.py --config=gpt2_configs/<config_file>

**Reply by BoxiangW** [reply]
Hi, yes. You can follow the [readme file](https://github.com/hpcaitech/ColossalAI-Examples/tree/main/language/gpt) to train gpt. You can set your own configurations based on your devices.

**Reply by gogogwwb** [reply]
Does multi-machine distributed training need to configure ssh password-free?

**Reply by FrankLeeeee** [reply]
Yes, this is required.

**Reply by gogogwwb** [reply]
When multi-machine distributed training, only one machine is running, why is this?Need to modify the code?[resnet](https://github.com/hpcaitech/ColossalAI-Examples/tree/main/image/resnet)，[pipeline_parallel](https://github.com/hpcaitech/ColossalAI-Examples/tree/main/features/pipeline_parallel)
![image](https://user-images.githubusercontent.com/51263129/174768264-7720466f-b0b3-46be-9f81-97f6fd2f9c47.png)

**Reply by gogogwwb** [reply]
other machine  seems to be running in the background.

**Reply by gogogwwb** [reply]
Is this a parameter setting incorrect?
![image](https://user-images.githubusercontent.com/51263129/174978580-654f5268-179b-4ccc-843a-51ef30316856.png)

**Reply by FrankLeeeee** [reply]
> When multi-machine distributed training, only one machine is running, why is this?Need to modify the code?[resnet](https://github.com/hpcaitech/ColossalAI-Examples/tree/main/image/resnet)，[pipeline_parallel](https://github.com/hpcaitech/ColossalAI-Examples/tree/main/features/pipeline_parallel) ![image](https://user-images.githubusercontent.com/51263129/174768264-7720466f-b0b3-46be-9f81-97f6fd2f9c47.png)

To do a simple check, you can write a simple test file like
```python
pritn('launched')
```

then run `colossalai run --nproc_per_node 2 --host host1,host2 --master_addr host1 test.py.

You should be able to see 4 lines of `launched`.

**Reply by FrankLeeeee** [reply]
> Is this a parameter setting incorrect? ![image](https://user-images.githubusercontent.com/51263129/174978580-654f5268-179b-4ccc-843a-51ef30316856.png)

For this one, seems that your gloo has some issue. You can check your network interface by `ifconfig` and look for the one for your ethernet. Lastly, run `GLOO_SOCKET_IFNAME=<interface-name> colossalai run ...`

**Reply by gogogwwb** [reply]
Is the reason for this error because of the config file?
![image](https://user-images.githubusercontent.com/51263129/175053786-d92c5342-4cda-41de-8f4c-5ba87d8acd5c.png)
![image](https://user-images.githubusercontent.com/51263129/175053889-e316281c-2066-4455-8cb5-fb546dddc44e.png)
